File: Vika9/InheritancePolymorphism.cpp
#include <iostream>
#include <string>
using namespace std;


class Vehicle {
    public:
        string make;
        string model;
    
        Vehicle(string make, string model) {
            this->make = make;
            this->model = model;
        }
    
        virtual string move() {
            return "Move!";
        }
    
        string str() {
            return make + " " + model;
        }

};

class Car : public Vehicle {
public:
    Car(string make, string model) : Vehicle(make, model){
    }

    string move() {
        return "Drive!";
    }
};

class Boat : public Vehicle {
public:
    Boat(string make, string model) : Vehicle(make, model){
    }
    
    string move() {
        return "Sail!";
    }
};

class Plane : public Vehicle {
public:
    Plane(string make, string model) : Vehicle(make, model){
    }
    
    string move() {
        return "Fly!";
    }
};


int main() {
    Vehicle vehicle("Generic", "Vehicle");
    Car car("Ford", "Mustang");
    Boat boat("Ibiza", "Touring 20");
    Plane plane("Boeing", "747");
    Vehicle* vehicles[] = {&vehicle, &car, &boat, &plane};
    for (int i = 0; i < 4; i++) {
        cout << vehicles[i]->str() << " " << vehicles[i]->move()<< endl;
    }
    return 0;
}

File: Vika9/InheritancePolymorphism.py

class Vehicle:
    def __init__(self, make, model):
        self.make = make
        self.model = model

    def move(self):
        return "Move!"

    def __str__(self):
        return f"{self.make} {self.model}"

class Car(Vehicle):
    def move(self):
        return "Drive!"

class Boat(Vehicle):
    def move(self):
        return "Sail!"

class Plane(Vehicle):
    def move(self):
        return "Fly!"

    

# Main program starts here
if __name__ == "__main__":
    vehicle = Vehicle("Generic", "Vehicle")
    car = Car("Ford", "Mustang")
    boat = Boat("Ibiza", "Touring 20")
    plane = Plane("Boeing", "747")
    for v in [vehicle, car, boat, plane]:
        print(str(v), v.move())


# Úttak / Ouput:
# Generic Vehicle Move!
# Ford Mustang Drive!
# Ibiza Touring 20 Sail!
# Boeing 747 Fly!